### PR TITLE
Payments: Synchronize the billing country dropdown between payment methods

### PR DESCRIPTION
**Pull request by DavidRothstein**
## Summary

This pull request is a small usability improvement which, code-wise, also lays the groundwork for https://github.com/Automattic/wp-calypso/issues/22033.

The usability improvement introduced here is that when a user is entering credit card or PayPal information and chooses a country from the dropdown, their future payment methods will now pre-select that country by default.

In other words, if you're adding a credit card and choose Germany as the billing country for that card, and you later decide to switch over to PayPal, you'll now see Germany automatically pre-selected on the PayPal form, which is extremely likely to be what you intend to choose there.  (Of course, if that guess is wrong, you still have the ability to change it.)

This happens at:
- The credit card and PayPal options on the checkout page, when you go to pay for something in your shopping cart.
- The page for adding a credit card (```/me/purchases/add-credit-card```).
- The page for editing an existing credit card (which you get to from ```/me/purchases```, then clicking on one of the purchases, then clicking "Edit Payment Method").

A country selection made at any of those pages will become the new default for all the others.

## Code changes

On the code level, this is accomplished by adding the payment country to the Redux state, and providing a new component called ```PaymentCountrySelect``` which wraps the existing ```CountrySelect``` component.  Any elements in the UI which want to use (and affect) the global payment country default can simply switch over to using this component.  This framework is what will be extended in later pull requests to achieve the rest of the goals in https://github.com/Automattic/wp-calypso/issues/22033.

## Other bugs fixed as a side effect

I needed to do a bit of extra work to make this pull request work correctly with EBANX payments for users located in Brazil (who have extra fields to fill out on the credit card form). As a side effect of doing that, this also fixes a couple (small) preexisting bugs with the EBANX payments:
1. If a Brazilian user chose Brazil on the credit card tab, then switched to the PayPal tab, then switched back, Brazil was selected, but the EBANX fields were no longer there. This is now fixed.
2. If a Brazilian user chose another country (for example, United States) on the credit card tab, then switched to Brazil, the phone number format for the extra EBANX fields defaulted to the first country (United States). It should actually default to Brazil, and only ever default to something else if the user intentionally changed it. This is now fixed.

In addition, this pull request is the first time that any of the fields on the credit card form were given a default value.  It turns out the existing code had bugs around the default value handling that needed to be fixed.  Those bugs were most prominently visible as a result of this pull request (because the payment country is now selected by default), but there actually were some more complicated ways to reproduce the bugs beforehand also, and those preexisting bugs are now fixed too. For example:
1. During checkout, if you filled out some fields on the credit card form, switched to PayPal, then switched back, when you tried to submit the credit card form your pre-filled values were treated as empty, thereby preventing you from submitting the form until you changed them.  This is now fixed.
2. If you had a stored card during checkout and switched back and forth between the stored card and the new card form, any credit card number you entered in the new card form before switching would be considered invalid (because the spaces in the card number shown on the screen, e.g. "4111 1111 1111...", were being erroneously saved to the credit card number in the payment data also). This is now fixed.

### CODE REVIEW

File: client/components/credit-card-form-fields/index.jsx [hunk 1]
```diff
@@ -98,28 +98,19 @@ export class CreditCardFormFields extends React.Component {
...
+		return isEbanxEnabledForCountry( this.getFieldValue( 'country' ) );
 	}
 
 	renderEbanxFields() {
 		const { translate, countriesList } = this.props;
-		const { countryCode, phoneCountryCode, countryName } = this.state;
+		const { userSelectedPhoneCountryCode } = this.state;
+		const countryCode = this.getFieldValue( 'country' );
+		const countryData = find( countriesList.get(), { code: countryCode } );
+		const countryName = countryData ? countryData.name : translate( 'this country' );
```
Comment (a8ci18n) [inline client/components/credit-card-form-fields/index.jsx]: :disappointed: 76 existing translations will be lost with this change.
If the string must change, the following option has already been translated **36** times:
`translate( 'Country', { context: 'Domain Edit Contact Info form.'} )` *ES Score: 9*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/sync-billing-country-across-payment-methods)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by DavidRothstein** [comment]
Note: This is mostly working correctly, but still "in progress" because:
- I am aware of one bug that can occur when you have a stored credit card but are paying with a new card instead (it's possible for the country to not be correctly stored and submitted in that case).  I still need to track down the cause of that.
- I still need to deal with writing/fixing automated tests (and do some more manual testing).

**Comment by DavidRothstein** [comment]
cc @Automattic/neutron (this mention probably won't work since I'm not a member of the team though)

File: client/components/credit-card-form-fields/index.jsx (hunk 1 above)
Comment (DavidRothstein) [reply client/components/credit-card-form-fields/index.jsx]: I am not sure what the i18n bot is saying here (it's a new string, not a changed string); however, I am also refactoring the code a bit to improve it so this extra string won't wind up being necessary anyway.

**Comment by DavidRothstein** [comment]
I pushed up a few new commits here to:
- Fix the failing tests
- Improve country validation and thereby remove the new/unnecessary translated string that the a8ci18n bot was complaining about
- Minor documentation improvements

Still to do:
- Write new tests.
- Fix the bug mentioned in an earlier comment (which can be reproduced by saving a stored credit card, then going through checkout and choosing the new credit card option rather than the stored card; when you do that and fill out the new credit card information, if you don't change the country dropdown from the default it thinks you haven't chosen a country and won't let you submit).
- More manual testing.

I will be back working on those things again probably sometime tomorrow (Friday).

**Comment by klimeryk** [comment]
Thanks for creating this PR and the work so far! 🙇 An extra improvement (for a future PR, perhaps) would be to re-use the country value from the [domain contact form](https://github.com/Automattic/wp-calypso/blob/613988fbdb64d25a3422cd2d7cc13310c4e6e296/client/my-sites/checkout/checkout/domain-details-form.jsx) (shown before the Payments form _if_ the user is registering a domain name as well) to make the process even smoother.

**Comment by DavidRothstein** [comment]
@klimeryk and @sirbrillig Thank you for the reviews!  Yes, there is a bit more work still to do here, but I'm glad to hear that I seem to be on the right track.

Regarding the domain contact form, yes, the plan is to include that in a future PR.  I wonder if it needs to be handled a little differently than the others though.  If someone is paying for the first time, then I agree it definitely makes sense to use the domain contact country as the default during checkout.  But if they've already paid for other things with a United States billing address and then buy a domain with a Canadian address for the contact info... I am not sure we want to assume that future billing addresses will be in Canada also.  But it's not a terrible assumption either, so I'm probably overthinking this :)

**Comment by klimeryk** [comment]
Yeah, historically we've had _very_ weird configurations and cases here, so I'd just start with the simplest case: if there's no prior payment info but there's domain contact info, pull the country from it (and maybe fallback to GeoIP otherwise? we already have the location of the user in Redux state FYI)

**Comment by DavidRothstein** [comment]
I pushed up some commits to fix the remaining bugs.  The last thing to do here is write some tests (which I'll do tomorrow) and then this should be ready for final review.

> (and maybe fallback to GeoIP otherwise? we already have the location of the user in Redux state FYI)

Yes, agreed - I'm also planning to propose that for a future PR.  It will become especially important later, because the country selection in the dropdown is going to start influencing other things on the page (cart currency, available payment methods, etc), so having something selected in that dropdown by default will make the effect of future selections more clear - and besides, it's better usability anyway.

**Comment by DavidRothstein** [comment]
I didn't wind up having much time to work on this yesterday so I was a bit delayed in getting it finished, but as of today the tests are now written and pushed.

I believe this pull request is therefore now ready for final review.

**Comment by DavidRothstein** [comment]
It looks like the canary tests timed out.  As far as I can tell from other PRs, this is a common occurrence (and unlikely to be anything caused by this PR)... Is there some way to retrigger them?  I'm not sure I have the right permissions to do that.

All the other tests passed, so I'm leaving this at "Needs Review".

**Comment by hoverduck** [comment]
@DavidRothstein removing and re-adding the Needs Review label will re-trigger the tests.  We just switched backends for calypso.live, and have been working through a few issues.

I'll go ahead and re-trigger the tests now so I can watch the canaries :)

**Comment by DavidRothstein** [comment]
Thanks @hoverduck!  I probably could have figured that trick out if I had bothered to think about it more :)

**Comment by DavidRothstein** [comment]
I'm bumping this so it can hopefully get a final review.  (Once this is merged, most of the other client-side changes needed for this project should just be small changes on top of this one, which is why I'm hoping to get this in soon as a baseline.)

Also, since this has been here a few days, I've updated this branch to include the latest from the master branch.  I also added the new/beta "Needs e2e Testing" tag to run a larger set of e2e tests against the branch, since it makes some decent-sized changes to the payment page behavior.

**Comment by DavidRothstein** [comment]
The full e2e tests are actually failing, with all the failures looking something like this:

```
02:34:28 FAILED: Taking screenshot of: 'https://calypso.live/checkout/e2eflowtesting1518834787179930.wordpress.com/business'
02:34:31   16 passing (undefinedms)
02:34:31   1 failing
02:34:31   1) [WPCOM] Sign Up  (desktop, en) Sign up for a site on a business paid plan w/ domain name coming in via /create as business flow @parallel Step 1: About Page Step 2: Themes Step 3: Domains Step 4: Account Step 5: Processing Step 6: Secure Payment Page Can enter and submit test payment details:
02:34:31      TimeoutError: Timed out waiting for element with css selector of '*[id="name"]' to be settable to: 'End To End Testing'
```

So it cannot find the credit card form (or at least can't find the name field on that form), which certainly could be related to this PR.  However, when I try it manually, everything works fine.  I'll have to look into this more closely next week, I guess.

**Comment by alisterscott** [comment]
@Automattic/flowpatrol @hoverduck I've been trying to work out why the sign up tests are failing, but I can't reproduce this locally on this branch. Another set of 👀 would be great to see if you can work out why?

**Comment by hoverduck** [comment]
It's bizarre, but seems to be related to something between WebDriver and older versions of Chrome.

The e2e tests for branches run against CircleCI 1.0, which is currently using Chrome v54.  Running the tests against that results in weird behavior on the `name` field of the payment form, where the text is typed in and immediately overwritten, never fully finishing.  Manually typing and copy/pasting into the field seems to be fine.

I manually (temporarily) installed Chrome v61 (the version we're currently using on CI 2.0) and the signup tests passed.

**Comment by DavidRothstein** [comment]
Wow, thanks for tracking that down!

For what it's worth, I think this code change (in ```client/components/credit-card-form-fields/index.jsx```) may be responsible for the behavior difference:

```
handleFieldChange = event => {
...
- this.setState( newState, () => this.updateFieldValues( fieldName, nextValue ) );
+ this.updateFieldValues( event.target.name, event.target.value );
 };
```

The old code led to an extra unnecessary rerender, and maybe that was inadvertently working around the Chrome problem in the tests by forcing the field to render with the correct value in the end.

What is the best strategy for addressing this?  I assume that in the long term, the solution is to update to a newer version of CircleCI, but I'm wondering what to do in the meantime... Do we need to hold off merging this PR to avoid breaking the tests?

If necessary, I could look at reverting the above change.  I only removed the setState() because after the other changes in this PR there was no longer any new state state to set, but it could probably be added back as a no-op (I'd have to test to be sure).  But it would be better not to have to do that.

**Comment by hoverduck** [comment]
I'm not sure the best way to approach this.  I've started a P2 discussion on it here - p6fDka-v0-p2

TL;DR - The tests run in Chrome, controlled by Chromedriver.  The CircleCI containers used in these tests have old versions of both.  Upgrading one or the other doesn't work well, we have to upgrade both, but I'd like to figure out the least impactful way to do that.

Do you have anything depending on merging this change?  Will it hurt to hold on it for another day or so while we figure it out?  If so, just let us know and we can make it work.

**Comment by DavidRothstein** [comment]
@hoverduck I will eventually have things that depend on merging this change, but not immediately.  Holding on to it for a day or so will definitely be fine, especially since I still need to get a final code review here before this is mergeable anyway.

Thank you for continuing to work on the problems with the tests.

**Comment by alisterscott** [comment]
Sorry for all the messages but the issue with chrome and the e2e automated tests should be resolved now, so there's no issue with this PR from that perspective. Thanks

**Comment by DavidRothstein** [comment]
Great, thanks to both of you!

A question about merging this:  My understanding is that people only merge their pull requests when they're going to be immediately deployed, but I'm pretty sure I don't have the permissions to do a deployment myself (I do seem to have the ability to merge).  So what's the best way to get this merged?  Have someone else merge/deploy it at a time when I'm around?

**Comment by alisterscott** [comment]
@DavidRothstein I can do this for you - I'll DM you now

**Comment by DavidRothstein** [comment]
Note for posterity: The followup for this is now at https://github.com/Automattic/wp-calypso/pull/23810 (although it still needs some polish and is also waiting on some WordPress.com REST API changes to get in).

The various followup tasks discussed above (regarding using the GeoIP country as default and integrating this with the domain contact form) should all be handled by that pull request.  In retrospect, I maybe should have split some of those off into separate pull requests (and I still could).  But I didn't want to maintain too many pull requests that touched the same area of the code.
